### ISSUE TITLE
NequIP-OAM-L + enable_OpenEquivariance produces non-conservative forces in production MD

### ISSUE THREAD

**Issue by iansteeg** [mir-group/nequip#610]
**Describe the bug**
NequIP-OAM-L-0.1 (and any l_max = 3 NequIP model), when run through nequip's
own `enable_OpenEquivariance` modifier, produces forces that are not the
gradient of the model's own energy. We measured this by finite differences
(Hessian-symmetry / curl F = 0 test) and confirmed it causally with an MD
counterfactual: the same packaged weights, differentiated with vs. without the
modifier, go from +79.7 eV/ns of spurious energy injection in NVE to a
residual consistent with zero. Root cause traced to OpenEquivariance's
kernels (filed upstream: https://github.com/PASSIONLab/OpenEquivariance/issues/211)
— the nequip integration path itself (`_tp_scatter_oeq.py`) is unaffected
between 0.18.0 and current main; the tensor-product kernel math is untouched
on nequip's side. Flagging here because `enable_OpenEquivariance` is nequip's
own documented modifier and NequIP-OAM-L is a nequip-distributed model — users
following nequip's docs to accelerate L-capacity models get non-conservative
forces with no warning. S, M and XL capacities are unaffected.

**To Reproduce**
Minimal nequip-facing code:

    from nequip.model import ModelFromPackage, modify

    def load(pkg, oeq):
        m = ModelFromPackage(pkg, compile_mode="eager")
        if oeq:
            m = modify(model=m, modifiers=[{"modifier": "enable_OpenEquivariance"}])
        return (next(iter(m.values())) if isinstance(m, dict) else m).eval()

    ref = load("NequIP-OAM-L-0.1.nequip.zip", oeq=False)
    acc = load("NequIP-OAM-L-0.1.nequip.zip", oeq=True)
    # forces from ref vs acc on the same structure differ by 1-4%, while the
    # two models' predicted ENERGY agrees to 5e-9 relative -- and acc's forces
    # fail a Hessian-symmetry (curl F = 0) check with an asymmetry that does
    # not shrink with the finite-difference step (a real defect, not
    # truncation error).

Full self-contained finite-difference reproducer (builds its own structure,
needs only a NequIP-OAM `.zip`) is inlined in
https://github.com/PASSIONLab/OpenEquivariance/issues/211

**Expected behavior**
`enable_OpenEquivariance` is documented as a pure inference-acceleration
modifier; we expected forces from the modified and unmodified model to agree
to floating-point precision, as they do for S, M and XL. For L they do not.

**Environment (please complete the following information):**
 - OS: Linux (Cray EX / LUMI-G, containerized LAMMPS+NequIP image)
 - python version: 3.12.13
 - python environment:
   - nequip version: 0.18.0
   - e3nn version: 0.6.0
   - pytorch version: 2.9.1+rocm6.3
 - GPU support: AMD Instinct MI250X (gfx90a) via ROCm/HIP, not CUDA
   - torch.version.cuda: None
   - torch.version.hip: 6.3.42134-a9a80e791
 - openequivariance version: 0.6.8 (dependency, not directly asked above but
   relevant here)

**Additional context**
Full measurement (finite-difference scan, per-species localization to
nitrogen/carbon, the MD counterfactual, and what we have not yet tested) is in
the upstream OpenEquivariance issue:
https://github.com/PASSIONLab/OpenEquivariance/issues/211

Possibly related: #586 reports erratic fine-tuning behavior with OEQ +
compile on this same model family and capacity. We don't have a compile
measurement of our own, so we've left a comment there rather than claim
identity with this bug.

Workaround: build the ML-IAP artifact with an empty `modifiers` list (drop
`enable_OpenEquivariance`) for L-capacity models — costs about 3.55x
throughput but restores conservative forces.

**Comment by kavanase** [comment]
Noting that I think I have now also seen this issue.

**Comment by asglover** [comment]
@kavanase I did a little digging into this issue and was not about to root cause it. Can you share the hardware you observed the issue on? Was it also AMD hardware? or do you see this on Nvidia hardware as well? I have not been able to test on AMD hardware.

**Comment by kavanase** [comment]
@asglover thanks for looking into this! I ran some tests and yes it seems like this is an AMD-specific issue. Specifically, I observe it with PyTorch 2.10 on AMD (Frontier), and not with PyTorch 2.8 on NVIDIA (FasRC). 

Summary from Claude of these analyses:
```
Both sets measure the same two things: the force difference between the OEQ model and the plain package model, and the residual between each model's analytic force and a finite-difference (FD) derivative of its own energy at h = 0.01 Å. Energies agree between OEQ and plain in both sets (about 1e-5 eV), so energies never show the issue; the forces do.

AMD (Frontier, MI250X, from the handoff; project test frames), meV/Å

┌───────┬─────┬─────────────────────────────────────┬────────────────────────┬─────────────────┬─────────────┐
│ frame │  N  │ OEQ vs plain force diff, mean / max │   FD residual plain    │ FD residual OEQ │ OEQ / plain │
├───────┼─────┼─────────────────────────────────────┼────────────────────────┼─────────────────┼─────────────┤
│ 0     │ 71  │ 4.6 / 19                            │ 0.34 (GPU), 0.21 (CPU) │ 3.3             │ ~10         │
├───────┼─────┼─────────────────────────────────────┼────────────────────────┼─────────────────┼─────────────┤
│ 1     │ 191 │ 59 / 240                            │ 0.53 (GPU), 0.47 (CPU) │ 66.4            │ ~125        │
└───────┴─────┴─────────────────────────────────────┴────────────────────────┴─────────────────┴─────────────┘

Mean OEQ vs plain force difference over the 8 strided test frames: 4.6, 5.4, 4.0, 1.0, 59, 9.6, 21, 4.0 meV/Å, unrelated to atom count. In the step-size sweep (h = 0.005 to 0.08, 36 components) the plain model's residual followed the pure FD error curve with a minimum of 0.35 meV/Å growing as h², while the OEQ residual stayed flat at 3.2 to 3.6 meV/Å. A flat residual means the analytic forces are not the gradient of the energy. M and XL agreed with their plain models to about 0.004 meV/Å.

NVIDIA (FASRC A100, generated frames), meV/Å

┌─────────────────────┬─────┬─────────────────────────────────────┬───────────────────┬─────────────────┬─────────────┐
│        frame        │  N  │ OEQ vs plain force diff, mean / max │ FD residual plain │ FD residual OEQ │ OEQ / plain │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ Si64_rattled        │ 64  │ 0.001 / 0.002                       │ 0.27              │ 0.25            │ 0.9         │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ NaCl63_vacancy      │ 63  │ 0.001 / 0.002                       │ 0.19              │ 0.18            │ 0.9         │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ Al2O3_corundum30    │ 30  │ 0.002 / 0.004                       │ 0.47              │ 0.55            │ 1.2         │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ Fe54_C_interstitial │ 55  │ 0.001 / 0.008                       │ 0.66              │ 0.52            │ 0.8         │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ NH3_H2O_box         │ 96  │ 0.001 / 0.004                       │ 2.12              │ 2.08            │ 1.0         │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ acetamide_box       │ 72  │ 0.001 / 0.005                       │ 1.39              │ 1.43            │ 1.0         │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ CuAl108_alloy       │ 108 │ 0.001 / 0.002                       │ 0.46              │ 0.45            │ 1.0         │
├─────────────────────┼─────┼─────────────────────────────────────┼───────────────────┼─────────────────┼─────────────┤
│ TiN64_strained      │ 64  │ 0.001 / 0.002                       │ 0.85              │ 0.83            │ 1.0         │
└─────────────────────┴─────┴─────────────────────────────────────┴───────────────────┴─────────────────┴─────────────┘

The same picture holds for M and XL on NVIDIA, and for L with torch 2.11: force differences of 0.001 to 0.002 meV/Å mean and at most 0.008 max, FD residual ratios of 0.8 to 1.2.
```

I'll try to run some quick debugging on AMD to see if I can narrow down the issue at all.

**Comment by kavanase** [comment]
The issue seems to be `OpenEquivariance`'s HIP post-processing, which rewrites `__syncwarp()` as `__threadfence_block()`, a fence rather than a barrier, which is then affecting the radial MLIP outputs.

From Claude:
**Root cause, found on Frontier.** Running the MI250X (each `TensorProductScatter`problem of L against e3nn + scatter on a random symmetric graph, 256 nodes, 5846 edges) isolates it immediately. For the 23-instruction L conv problem in float32, non-deterministic mode, forward and the gradients w.r.t. node features and spherical harmonics are exact (2 to 3e-7 relative), but the gradient w.r.t. the edge weights has relative error 0.53, on 100 % of edges, and it comes from exactly one instruction:

```
ins16  32x2e x 1x3o -> 32x3o  uvu   dW err = 5.3e-01
all 22 others                        dW err < 2e-07
```

Edge weights are the output of the radial MLP of the interatomic distances, so a wrong weight gradient corrupts forces and leaves energies untouched, which is the whole symptom. The other two L problems (4 instructions each), all M and XL problems, and float64 are fine, and deterministic mode is exact for this problem too. (For location: on MI250X ins16 is the last instruction of backward segment 5 of 7.) NVIDIA passes because the substitution below is only applied under `torch.version.hip`; CUDA keeps a real `__syncwarp()`.

The trigger is the HIP kernel post-processing in `openequivariance/_torch/extlib/__init__.py`:

```python
kernel = kernel.replace("__syncwarp();", "__threadfence_block();")
```

`__threadfence_block()` is a memory fence, not an execution barrier. What I can show is that adding a wavefront barrier, a compiler scheduling barrier with no ISA cost, fixes the layer test completely, while keeping the safe `atomicAdd` changes nothing; that is consistent with the compiler reordering shared-memory accesses across the fence in one unrolled backward segment. Which segment gets reordered is then a scheduling accident of hiprtc on gfx90a for this particular code, which would be why the same templates are fine on CUDA, for M and XL, and for the other 22 instructions of L:

| HIP post-processing of `__syncwarp();` | `atomicAdd` | dW rel. err (ins16 problem) |
|---|---|---|
| `__threadfence_block();` (stock) | `unsafeAtomicAdd` (stock) | 5.3e-01 |
| `__threadfence_block();` | `atomicAdd` (safe) | 5.3e-01 |
| `__threadfence_block(); __builtin_amdgcn_wave_barrier();` | `unsafeAtomicAdd` | 1.8e-07 |
| `__builtin_amdgcn_wave_barrier();` | `unsafeAtomicAdd` | 1.8e-07 |

End to end, with the fence-plus-barrier substitution monkey-patched in before the OEQ-modified L model is built, the OEQ forces are identical to the plain e3nn path (max |ΔF| < 0.05 meV/Å on the three frames that were worst before, against 19 to 238 meV/Å stock), the finite-difference residual drops to 0.4 to 0.7 meV/Å (the plain model's level), M and XL stay exact, and the force call is no slower (92 ms vs 99 ms on a 191-atom cell). So the proposed upstream fix is a one-line change in `_torch/extlib/__init__.py` (and the same line in `jax/extlib/__init__.py`), keeping the fence for its memory-ordering semantics and adding the barrier:

```python
kernel = kernel.replace("__syncwarp();", "__threadfence_block(); __builtin_amdgcn_wave_barrier();")
```

(The barrier alone also passed every test; the combined form is the conservative choice.)

Happy to open the PR on PASSIONLab/OpenEquivariance if that is easiest. Until it lands, a ROCm user can apply the same substitution in Python before the model is built:

```python
import openequivariance._torch.TensorProductConv as TPC
_pp = TPC.postprocess_kernel
TPC.postprocess_kernel = lambda k: _pp(k).replace("__threadfence_block();", "__threadfence_block(); __builtin_amdgcn_wave_barrier();")
```

One unrelated observation from the generated kernel, for the OEQ side: the Clebsch–Gordan coefficients appear to be emitted at float32 precision (`1.08654`, `1.1751`), so float64 accuracy of the 23-instruction problem is capped at ~3e-8 relative while the scalar-only problems reach 1e-16. Harmless for float32 models, but worth knowing for anyone validating in double.

Test file here:
[oeq_layer_test.py](https://github.com/user-attachments/files/31748659/oeq_layer_test.py)

Output on the MI250X, stock, for reference:

ninst= 4 float32 atomic fwd=3.9e-07 dX=1.5e-07 dSH=1.8e-07 dW=1.6e-07
ninst=23 float32 atomic fwd=2.3e-07 dX=3.2e-07 dSH=1.6e-07 dW=5.3e-01
   ins16 32x2e    x 1x3o  -> 32x3o    uvu err=5.3e-01  <--    (all 22 others < 2e-07)
ninst=23 float32 det    fwd=1.9e-07 dX=3.7e-07 dSH=2.0e-07 dW=1.6e-07
ninst= 4 float32 atomic fwd=1.8e-07 dX=1.9e-07 dSH=1.4e-07 dW=7.6e-08

With --fix, the 23-instruction row reads dW=1.6e-07.

**Comment by asglover** [comment]
Sean, this was amazingly helpful, thank you so much! and thanks for the additional pointer to the CG coefficient literals being truncated for FP64. I'll look into that separately. And big congratulations on the assistant professorship! Great to see! Thanks again! Vivek is going to reproduce the fix, then we should have it merged asap!

**Comment by kavanase** [comment]
No problem! Glad it's helpful 😃  And thanks! 🙌

**Comment by asglover** [comment]
"One unrelated observation from the generated kernel, for the OEQ side: the Clebsch–Gordan coefficients appear to be emitted at float32 precision (1.08654, 1.1751), so float64 accuracy of the 23-instruction problem is capped at ~3e-8 relative while the scalar-only problems reach 1e-16. " I was not able to reproduce this or identify a section of code to be suspicious of. Could you help me reproduce this? how did you test or find this? Is this something you have consistently seen?

**Comment by kavanase** [comment]
@asglover thanks for fixing this! I've pulled the changes and confirmed it fixes the issue on Frontier.

I re-checked, and that `float64` remark was wrong, so please ignore. Apologies! The ~3e-8 came from taking the e3nn `TensorProduct` out of the (float32) nequip model and upcasting to double, which carried float32 Clebsch–Gordan buffers along. Against a e3nn `TensorProduct` built with `float64`, OEQ's `float64` output agrees to 5e-16. So all good!

**Comment by kavanase** [comment]
@iansteeg thanks again for raising!
@cw-tan this can be closed now (I don't have permissions)

**Comment by cw-tan** [comment]
closed by https://github.com/PASSIONLab/OpenEquivariance/pull/217
